mpas_analysis/shared/interpolation/interpolate.py

```diff
@@ -14,216 +14,521 @@
+        self.src_grid_dims = dsMapping['src_grid_dims'].values[::-1]
+        self.dst_grid_dims = dsMapping['dst_grid_dims'].values[::-1]
+
+        self.frac_b = dsMapping['frac_b'].values
+
+        col = dsMapping['col'].values-1
+        row = dsMapping['row'].values-1
+        S = dsMapping['S'].values
+        self.matrix = csr_matrix((S, (row, col)), shape=(n_b, n_a))
+
+        self.mappingLoaded = True  # }}}
+
+    def _check_drop(self, dataArray):  # {{{
+        sourceDims = self.sourceDescriptor.dims
+
+        sourceDimsInArray = [dim in dataArray.dims for dim in sourceDims]
+
+        return (numpy.any(sourceDimsInArray) and not
+                numpy.all(sourceDimsInArray))  # }}}
+
+    def _remap_data_array(self, dataArray, renormalizationThreshold):  # {{{
+        '''
+        Regrids a single xarray data array
+
+        Author
+        ------
+        Xylar Asay-Davis
+
+        Last Modified
+        -------------
+        04/05/2017
+        '''
+
+        sourceDims = self.sourceDescriptor.dims
+        destDims = self.destinationDescriptor.dims
+
+        sourceDimsInArray = [dim in dataArray.dims for dim in sourceDims]
+
+        if not numpy.any(sourceDimsInArray):
+            # no remapping is needed
+            return dataArray
+
+        if not numpy.all(sourceDimsInArray):
+            # no remapping is possible so the variable array should have been
+            # dropped
+            raise ValueError('Data array with some (but not all) required '
+                             'source dims cannot be remapped\n'
+                             'and should have been dropped.')
+
+        # make a list of dims and remapAxes
+        dims = []
+        remapAxes = []
+        destDimsAdded = False
+        for index, dim in enumerate(dataArray.dims):
+            if dim in sourceDims:
+                remapAxes.append(index)
+                if not destDimsAdded:
+                    dims.extend(destDims)
+                    destDimsAdded = True
+            else:
+                dims.append(dim)
+
+        # make a dict of coords
+        coordDict = {}
+        # copy unmodified coords
+        for coord in dataArray.coords:
+            sourceDimInCoord = numpy.any([dim in dataArray.coords[coord].dims
+                                          for dim in sourceDims])
+            if not sourceDimInCoord:
+                coordDict[coord] = {'dims': dataArray.coords[coord].dims,
+                                    'data': dataArray.coords[coord].values}
+
+        # add dest coords
+        coordDict.update(self.destinationDescriptor.coords)
+
+        # remap the values
+        field = dataArray.values
+        mask = numpy.isnan(field)
+        if numpy.count_nonzero(mask) > 0:
+            field = numpy.ma.masked_array(field, mask)
+        remappedField = self._remap_numpy_array(field, remapAxes,
+                                                renormalizationThreshold)
```

Comment:
Here, I convert the entire contents of a data array to a numpy array.  Based on memory constraints, it might be necessary to divide the array into slices along non-horizontal axes.  It *could* even be possible to divide along horizontal axes but that would be less than ideal.

I'm hesitant for now to hurt performance by slicing data arrays here.  Presumably if we run into trouble, we could always manually call remapping slice by slice on a big data set, passing data arrays for each slice and reassembling the full data set.  It may be best not to build that into the low level behavior.